### PR TITLE
Add error metrics and DLQ support to GCS/File write schema transforms

### PR DESCRIPTION
**Pull request by nickuncaged1201**
Add metrics to all `FileWriteSchemaTransforms`. A few refactoring is needed to make this possible
* Change all FileSchemaTransformFormatProvider to returning `PCollectionTuple` instead of just `PCollection<String>` to plumb error rows out
* Add an `ErrorCounterFn` that can be used generally with all FileSchemaTransforms.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by nickuncaged1201** [comment]
r: @Dippatel98 @manavgarg @johnjcasey

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformProvider.java [hunk 1]
```diff
@@ -129,7 +136,11 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
                                   .build()))
               .setRowSchema(OUTPUT_SCHEMA);
 
-      return PCollectionRowTuple.of(OUTPUT_TAG, output);
+      if (files.has(ERROR_TAG)) {
+        return PCollectionRowTuple.of(OUTPUT_TAG, output).and("error", files.get(ERROR_TAG));
```
Comment (manavgarg) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformProvider.java:140]: can we have a constant for "error" similar to OUTPUT_TAG. Let's change it all the places. Using hardcoded strings is not a good practice and makes refactoring difficult.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/JsonWriteSchemaTransformFormatProvider.java [hunk 2]
```diff
@@ -51,22 +60,33 @@ public String identifier() {
    * {@link PCollection} file names written using {@link TextIO.Write}.
    */
   @Override
-  public PTransform<PCollection<Row>, PCollection<String>> buildTransform(
+  public PTransform<PCollection<Row>, PCollectionTuple> buildTransform(
```
Comment (manavgarg) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/JsonWriteSchemaTransformFormatProvider.java:63]: nit: please fix the comments at all places where we change the method args.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/ParquetWriteSchemaTransformFormatProvider.java [hunk 3]
```diff
@@ -69,14 +80,25 @@ public PCollection<String> expand(PCollection<Row> input) {
 
         write = applyCommonFileIOWriteFeatures(write, configuration);
 
-        return input
-            .apply(
+        PCollectionTuple avro =
```
Comment (manavgarg) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/ParquetWriteSchemaTransformFormatProvider.java:83]: change avro to parquet.

**Review by manavgarg — COMMENTED** [review]
I don't think existing test changes are enough. Can we add more tests to test all the formats and also the scenario for checking for errored out records?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java [hunk 4]
```diff
@@ -47,22 +57,27 @@ public String identifier() {
...
-                .setCoder(coder);
+        PCollectionTuple tuple =
+            input.apply(
+                "Row To Avro Generic Record",
+                ParDo.of(
+                        new ErrorCounterFn<GenericRecord>(
+                            "Avro-write-error-counter",
+                            AvroUtils.getRowToGenericRecordFunction(AvroUtils.toAvroSchema(schema)),
+                            ERROR_FN_OUPUT_TAG))
+                    .withOutputTags(ERROR_FN_OUPUT_TAG, TupleTagList.of(ERROR_TAG)));
```
[the remark this answers is not in the document]
Comment (johnjcasey) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java:78]: I'm a bit confused by this. How is it counting errors upstream of where you are writing to avro?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java [hunk 5]
```diff
@@ -61,6 +73,49 @@ static MapElements<Row, GenericRecord> mapRowsToGenericRecords(Schema beamSchema
...
+      this.outputTag = outputTag;
+    }
+
+    @ProcessElement
+    public void process(@DoFn.Element Row row, MultiOutputReceiver receiver) {
+      try {
+        receiver.get(outputTag).output(mapFn.apply(row));
+      } catch (Exception e) {
+        errorsInBundle += 1;
+        LOG.warn("Error while parsing input element", e);
```
Comment (johnjcasey) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:102]: is this error always going to be valid? Id expect some mapper functions to not be parsers

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java [hunk 6]
```diff
@@ -61,6 +73,49 @@ static MapElements<Row, GenericRecord> mapRowsToGenericRecords(Schema beamSchema
...
   }
 
+  // Applies generic mapping from Beam row to other data types through the provided mapFn.
+  // Implemenets error handling with metrics and DLQ support.
+  // Arguments:
+  //    name: the metric name to use.
+  //    mapFn: the mapping function for mapping from Beam row to other data types.
+  //    outputTag: TupleTag for output. Used to direct output to correct output source, or in the
+  //        case of error, a DLQ.
+  static class ErrorCounterFn<OutputT extends Object> extends DoFn<Row, OutputT> {
```
Comment (johnjcasey) [inline sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:83]: this name doesn't seem right. This transform is a .map() that aggregates errors on a separate tag

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java:78]: r: @manavgarg The ERROR_FN_OUTPUT_TAG is for successful outputs. Let me know if there's a better name for this. 
r: @johnjcasey The AvroWriteIo doesn't expose any errors by itself, so I won't be able to count any error there. I implement the error counting for this conversion step to GenericRecords with the intention that if such a conversion fails, then definitely some error happens with the record. If the conversion is successful, I don't see a reason the write will fail anyway. This is a bit of an unfortunate misnomer to call it write level error count

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/CsvWriteSchemaTransformFormatProvider.java [hunk 7]
```diff
@@ -83,9 +85,11 @@ public PCollection<String> expand(PCollection<Row> input) {
...
 
         WriteFilesResult<String> result = input.apply("Row to CSV", write);
-        return result
-            .getPerDestinationOutputFilenames()
-            .apply("perDestinationOutputFilenames", Values.create());
+        PCollection<String> output =
+            result
+                .getPerDestinationOutputFilenames()
+                .apply("perDestinationOutputFilenames", Values.create());
+        return PCollectionTuple.of(RESULT_TAG, output);
```
[the remark this answers is not in the document]
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/CsvWriteSchemaTransformFormatProvider.java:92]: As in the previous note, CsvWriteIO itself doesn't expose any error. All other transforms have some sort of conversion step, while CSV write doesn't. So nothing can be done about this really.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 5 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:102]: For the case of file write transforms, all of the mapper functions are some sort of conversion parser from Beam row, be it converting to Avro generic records or Json etc. So I think for the scope this should be valid

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 6 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:83]: What would be a better name in this case? BeamRowMapperWithDlq something like this?

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/JsonWriteSchemaTransformFormatProvider.java (hunk 2 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/JsonWriteSchemaTransformFormatProvider.java:63]: Done

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformProvider.java (hunk 1 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformProvider.java]: Done

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProvider.java [hunk 8]
```diff
@@ -38,6 +39,6 @@ public interface FileWriteSchemaTransformFormatProvider extends Providers.Identi
    * Builds a {@link PTransform} that writes a {@link Row} {@link PCollection} and outputs the
    * resulting {@link PCollection} of the file names.
    */
-  PTransform<PCollection<Row>, PCollection<String>> buildTransform(
+  PTransform<PCollection<Row>, PCollectionTuple> buildTransform(
```
[the remark this answers is not in the document]
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProvider.java:42]: Done

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 6 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:83]: Done. Tests have been added for the class.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnjcasey) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/AvroWriteSchemaTransformFormatProvider.java:78]: got it, that makes sense

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 6 above)
Comment (johnjcasey) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:83]: error handling beam row mapper maybe? that or your suggestion works

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 5 above)
Comment (johnjcasey) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:102]: thats fair

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 6 above)
Comment (johnjcasey) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java:83]: also, ideally we would have a unit test for this isolated from any particular format. I do appreciate that you've added tests that use this for all the formats.

**Review by johnjcasey — CHANGES_REQUESTED** [review]

**Comment by manavgarg** [comment]
Thanks Nick for the changes. Please address the remaining comments from John but otherwise LGTM.

**Review by manavgarg — APPROVED** [review]
LGTM, after addressing John's remaining comments.

File: sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java (hunk 6 above)
Comment (nickuncaged1201) [reply sdks/java/io/file-schema-transform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformFormatProviders.java]: Done. Changed name to `BeamRowMapperWithDlq`, and added a generic test independent of any format.

**Comment by johnjcasey** [comment]
lgtm

**Pull request merged**